wrappers/pkcs11/pkcs11_client.go

```diff
@@ -0,0 +1,632 @@
+		keyLabel = api.ReadBaoVariable(EnvHsmWrapperKeyLabel)
+	case opts.withKeyLabel != "":
+		keyLabel = opts.withKeyLabel
+	default:
+		return nil, nil, fmt.Errorf("key label is required")
+	}
+
+	switch {
+	case api.ReadBaoVariable(EnvHsmWrapperMechanism) != "" && !opts.Options.WithDisallowEnvVars:
+		mechanism = api.ReadBaoVariable(EnvHsmWrapperMechanism)
+	case opts.withMechanism != "":
+		mechanism = opts.withMechanism
+	default:
+		mechanism = ""
+	}
+
+	switch {
+	case api.ReadBaoVariable(EnvHsmWrapperRsaOaepHash) != "" && !opts.Options.WithDisallowEnvVars:
+		rsaOaepHash = strings.ToLower(api.ReadBaoVariable(EnvHsmWrapperRsaOaepHash))
+	case opts.withRsaOaepHash != "":
+		rsaOaepHash = strings.ToLower(opts.withRsaOaepHash)
+	default:
+		rsaOaepHash = ""
+	}
+
+	if slot != "" {
+		if slotNum, err = numberAutoParse(slot, 32); err != nil {
+			return nil, nil, fmt.Errorf("Invalid slot number")
+		}
+	} else {
+		slotNum = 0
+	}
+
+	if mechanism != "" {
+		if mechanismNum, err = MechanismFromString(mechanism); err != nil {
+			return nil, nil, err
+		}
+	} else {
+		mechanismNum = 0
+	}
+
+	client := &Pkcs11Client{
+		client:      nil,
+		lib:         lib,
+		slot:        uint(slotNum),
+		pin:         pin,
+		tokenLabel:  tokenLabel,
+		keyId:       keyId,
+		keyLabel:    keyLabel,
+		mechanism:   uint(mechanismNum),
+		rsaOaepHash: rsaOaepHash,
+	}
+	
+	// Initialize the client
+	err = client.InitializeClient()
+	if err != nil {
+		return nil, nil, err
+	}
+	// Validate credentials for session establishment
+	session, err := client.GetSession()
+	if err != nil {
+		return nil, nil, err
+	}
+	defer client.CloseSession(session)
+
+	wrapConfig := new(wrapping.WrapperConfig)
+	wrapConfig.Metadata = make(map[string]string)
+	wrapConfig.Metadata["lib"] = lib
+	wrapConfig.Metadata["key_label"] = keyLabel
+	wrapConfig.Metadata["key_id"] = keyId
+	if slotNum != 0 {
+		wrapConfig.Metadata["slot"] = strconv.Itoa(int(slotNum))
+	}
+	if tokenLabel != "" {
+		wrapConfig.Metadata["token_label"] = tokenLabel
+	}
+	if mechanismNum != 0 {
+		wrapConfig.Metadata["mechanism"] = MechanismString(uint(mechanismNum))
+	}
+	if rsaOaepHash != "" {
+		wrapConfig.Metadata["rsa_oaep_hash"] = rsaOaepHash
+	}
+
+	return client, wrapConfig, nil
+}
+
+
+func (c *Pkcs11Client) Close() {
+	if c.client == nil {
+		return
+	}
+	c.client.Finalize()
+	c.client.Destroy()
+	c.client = nil
+}
+
+func (c *Pkcs11Client) Encrypt(plaintext []byte) ([]byte, *Pkcs11Key, error) {
+	session, err := c.GetSession()
+	if err != nil {
+		return nil, nil, err
+	}
+	defer c.CloseSession(session)
+
+	keyId := Pkcs11Key{ label: c.keyLabel, id: c.keyId }
+	key, err := c.FindKey(session, keyId, pkcs11.CKA_ENCRYPT)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	prefix, params, err := c.CreatePkcsParamsForKey(session, key)
+	if err != nil {
+		return nil, nil, err
+	}
+	if err = c.client.EncryptInit(session, []*pkcs11.Mechanism{params}, key); err != nil {
```

Comment:
@glatigny Hmm, this currently panics with AES-GCM and the Dockerfile above... 

```
/usr/include/c++/14/bits/stl_vector.h:1127: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = unsigned char; _Alloc = SecureAllocator<unsigned char>; reference = unsigned char&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
SIGABRT: abort
PC=0x7f83f2642c94 m=0 sigcode=18446744073709551610
signal arrived during cgo execution

goroutine 8 gp=0xc00018e8c0 m=0 mp=0xae6780 [syscall]:
runtime.cgocall(0x734cf0, 0xc000086c38)
	/usr/lib/golang/src/runtime/cgocall.go:167 +0x4b fp=0xc000086c10 sp=0xc000086bd8 pc=0x46bb2b
github.com/miekg/pkcs11._Cfunc_EncryptInit(0x3ef3bf50, 0x2, 0xc0000123f0, 0x3)
	_cgo_gotypes.go:873 +0x4c fp=0xc000086c38 sp=0xc000086c10 pc=0x719ccc
github.com/miekg/pkcs11.(*Ctx).EncryptInit.func1(0xc000086dd0?, 0x2, 0xc0000123f0, 0x3)
	/root/go/pkg/mod/github.com/miekg/pkcs11@v1.1.2-0.20231115102856-9078ad6b9d4b/pkcs11.go:1155 +0x7d fp=0xc000086c78 sp=0xc000086c38 pc=0x71d77d
github.com/miekg/pkcs11.(*Ctx).EncryptInit(0xc000072220, 0x2, {0xc000086dd0?, 0x7f83f2596748?, 0x0?}, 0x3)
	/root/go/pkg/mod/github.com/miekg/pkcs11@v1.1.2-0.20231115102856-9078ad6b9d4b/pkcs11.go:1155 +0x99 fp=0xc000086ce8 sp=0xc000086c78 pc=0x71d659
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.(*Pkcs11Client).Encrypt(0xc0000e5780, {0xc0001cb978, 0x3, 0x3})
	/go-kms-wrapping/wrappers/pkcs11/pkcs11_client.go:249 +0x1a5 fp=0xc000086df8 sp=0xc000086ce8 pc=0x72f205
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.(*Wrapper).Encrypt(0xc000086f08, {0x762220?, 0x87d070?}, {0xc0001cb978?, 0x50a200?, 0xa2ccd8?}, {0xaaff60?, 0xc0001cb8a8?, 0x50a22d?})
	/go-kms-wrapping/wrappers/pkcs11/pkcs11.go:76 +0x43 fp=0xc000086e90 sp=0xc000086df8 pc=0x72dc63
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.TestAccPkcs11Wrapper_Lifecycle(0xc0000dcea0)
	/go-kms-wrapping/wrappers/pkcs11/pkcs11_acc_test.go:36 +0x155 fp=0xc000086f70 sp=0xc000086e90 pc=0x7318f5
testing.tRunner(0xc0000dcea0, 0x8197c0)
	/usr/lib/golang/src/testing/testing.go:1690 +0xf4 fp=0xc000086fc0 sp=0xc000086f70 pc=0x50a2f4
```

Reply:
Okay, I will rollback. I wanted to avoid returning a `[]*pkcs11.Mechanism`., but that's not a safe choice..
That's strange that it didn't failed on my laptop.

Reply:
Sorry, see the comment below, this was crashing because my SoftHSM on Fedora 40 didn't like the 16-byte prefix for GCM. It didn't have to do with this.

Reply:
According to Oasis Documentation, we can safely use a 12 bytes length for the nonce:
https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/pkcs11-curr-v3.0.html#_Toc228894714

Since GCM has some specific behavior (free params, IV can be added by the HSM...) ; I'm performing some refactoring.
I'll see if I can keep some common code but it's possible that I'll split AES and RSA encrypt/decrypt in dedicated functions.